mitiq/zne/scaling/folding.py

```diff
@@ -421,6 +289,262 @@ def fold_all(
+
+    # If the scale_factor is an odd integer, we are done.
+    if np.isclose(odd_integer_scale_factor, scale_factor):
+
+        return num_folds_mask
+
+    # If necessary, fold a subset of gates to approximate the scale_factor
+    input_circuit_weight = sum(weight_mask)
+    output_circuit_weight = odd_integer_scale_factor * input_circuit_weight
+
+    # Express folding order through a list of indices
+    folding_order = list(range(len(weight_mask)))
+    if folding_method == "from_left":
+        pass
+    elif folding_method == "from_right":
+        folding_order.reverse()
+    elif folding_method == "at_random":
+        rnd_state = np.random.RandomState(seed)
+        rnd_state.shuffle(folding_order)
+    else:
+        raise ValueError(
+            "The option 'folding_method' is not valid."
+            "It must be 'at_random', or 'from_left', or 'from_right'."
+        )
+
+    # Fold gates until the input scale_factor is better approximated
+    approx_error = np.abs(
+        output_circuit_weight - scale_factor * input_circuit_weight
+    )
+    for j in folding_order:
+        # Skip gates with 0 weight
+        if np.isclose(weight_mask[j], 0.0):
+            continue
+
+        # Compute the approx error if a new fold would be applied
+        new_output_circuit_weight = output_circuit_weight + 2 * weight_mask[j]
+        new_approx_error = np.abs(
+            new_output_circuit_weight - scale_factor * input_circuit_weight
+        )
+        # Fold the candidate gate only if it helps improving the approximation
+        if new_approx_error < approx_error:
+            approx_error = new_approx_error
+            output_circuit_weight = new_output_circuit_weight
+            num_folds_mask[j] += 1
+        # TODO: decide if removing next else case to get better approximations
+        else:
+            break
+
+    return num_folds_mask
+
+
+def _apply_fold_mask(
+    circuit: Circuit,
+    num_folds_mask: List[int],
+    squash_moments: Optional[bool] = True,
+) -> Circuit:
+    r"""Applies local unitary folding to the gates of the input circuit
+    according to the input num_folds_mask.
+
+    More precicely, G_j^\dag G_j is applied after the j_th gate G_j of
+    the input circuit an integer number of times given by num_folds_mask[j].
+
+    Args:
+        circuit: The quantum circuit to fold.
+        num_folds_mask: The list of integers indicating how many times
+            the corresponding gates of 'circuit' should be folded.
+        squash_moments: If True or None, all gates (including folded gates) are
+            placed as early as possible in the circuit. If False, new moments
+            are created for folded gates. This option only applies to QPROGRAM
+            types which have a "moment" or "time" structure. Default is True.
+
+    Returns: The folded quantum circuit.
+    """
+    if squash_moments is False:
+        pass
+    else:
+        squash_moments = True
```

Comment:
huh?

Reply:
it was a very ugly way of setting `squash_moments = True` when `squash_moments = None`.  Now I removed this and used the default option of the `kwargs.get() ` method instead.  ( 9bb5e16)